File: src/zig/diff.zig
// Diffing algorithm for efficient updates
const std = @import("std");
const Allocator = std.mem.Allocator;

pub const DiffOp = union(enum) {
    no_change: void,
    update_line: struct { line: u32, content: []u8 },
    delete_line: u32,
    insert_line: struct { line: u32, content: []u8 },
};

pub fn diff_frames(
    prev: [][]u8,
    curr: [][]u8,
    allocator: Allocator,
) ![]DiffOp {
    var ops = std.ArrayList(DiffOp){};
    errdefer ops.deinit(allocator);

    const max_len = @max(prev.len, curr.len);

    for (0..max_len) |i| {
        const prev_line = if (i < prev.len) prev[i] else null;
        const curr_line = if (i < curr.len) curr[i] else null;

        if (prev_line == null and curr_line != null) {
            // Line inserted
            try ops.append(allocator, .{ .insert_line = .{ .line = @intCast(i), .content = curr_line.? } });
        } else if (prev_line != null and curr_line == null) {
            // Line deleted
            try ops.append(allocator, .{ .delete_line = @intCast(i) });
        } else if (prev_line != null and curr_line != null) {
            // Check if line changed
            if (!std.mem.eql(u8, prev_line.?, curr_line.?)) {
                try ops.append(allocator, .{ .update_line = .{ .line = @intCast(i), .content = curr_line.? } });
            } else {
                try ops.append(allocator, .no_change);
            }
        }
    }

    return try ops.toOwnedSlice(allocator);
}

File: src/native/diff.ts
// Diffing algorithm for efficient updates - TypeScript implementation
// Optimized for Node.js performance

export type DiffOp =
  | { type: 'no_change' }
  | { type: 'update_line'; line: number; content: string }
  | { type: 'delete_line'; line: number }
  | { type: 'insert_line'; line: number; content: string };


/**
 * Diff two frames to generate minimal update operations.
 * 
 * Strategy:
 * - If only a small continuous region of a line changes, use partial update
 * - If multiple regions change or changes are scattered, redraw the whole line
 * - When lines are deleted, all lines below need to be redrawn (reflow)
 */
export function diffFrames(prev: string[], curr: string[]): DiffOp[] {
  const ops: DiffOp[] = [];
  const maxLen = Math.max(prev.length, curr.length);
  let needsReflow = false; // Track if we need to redraw lines below

  for (let i = 0; i < maxLen; i++) {
    const prevLine = i < prev.length ? prev[i] : null;
    const currLine = i < curr.length ? curr[i] : null;

    if (prevLine === null && currLine !== null) {
      // Line inserted - causes reflow
      needsReflow = true;
      ops.push({ type: 'insert_line', line: i, content: currLine });
    } else if (prevLine !== null && currLine === null) {
      // Line deleted - causes reflow of everything below
      needsReflow = true;
      ops.push({ type: 'delete_line', line: i });
    } else if (prevLine !== null && currLine !== null) {
      if (prevLine === currLine) {
        // No change, but if we had reflow above, we need to redraw
        if (needsReflow) {
          ops.push({ type: 'update_line', line: i, content: currLine });
        } else {
          ops.push({ type: 'no_change' });
        }
      } else {
        // Line changed
        // For now, we always redraw the whole line for simplicity
        // Partial updates with ANSI codes are complex and error-prone
        // The optimization would be: if change is small and continuous, update just that region
        // But preserving ANSI codes at boundaries is tricky, so we redraw for correctness
        needsReflow = true; // Any line change causes reflow below
        ops.push({ type: 'update_line', line: i, content: currLine });
      }
    }
  }

  return ops;
}

